### ISSUE TITLE
Question about connection

### ISSUE THREAD

**Issue by esegas** [axopy/pytrigno#2]
Hi,
I'm trying to use your program in order to read data recorded by the trigno EMG and IMU system.
I keep raise de "Device disconnected" error. I think it is because I don't have the good samples_per_read value.
I used 200 because the sample rate is 2 kHz for EMG and I saw in the other question that the update rate is 10 Hz. But, I'm not sure what this update rate refers to.
Can you help me ?

Best regards,
Effie

**Comment by ixjlyons** [comment]
Could you provide a minimal example that produces the error, or are you using the example code from #1?

If you're getting that error, it should mean you've connected to the Trigno Control Utility socket(s), but could you specify the ports and IP address listed in TCU's settings?

In my experience, it shouldn't matter how many samples you request per read, it should just work (as opposed to some other data acquisition devices that require the number of samples requested to be a multiple of 512 bytes or something like that). With a 2 kHz sample rate, asking for 200 samples (per channel) would take 100 ms to fill the buffer, so if you repeatedly request samples, you'll see updates come in 10 times per second.

**Comment by ixjlyons** [comment]
Closing due to inactivity. If you're still experiencing issues connecting, let me know and we can re-open.

**Comment by jdekarske** [comment]
Hi Kenny,
Having the same issue, the ports in TCU match pytrigno.py and I'm using localhost. I updated the base and sensor firmware, but I can't imagine that would be the problem. Here's my output:
```(powersenv) C:\Users\emg\Documents\powers_Jason\pytrigno>c:/Users/emg/Documents/powers_Jason/powersenv/Scripts/python.exe c:/Users/emg/Documents/powers_Jason/pytrigno/examples/check_trigno.py
Traceback (most recent call last):
  File "c:\Users\emg\Documents\powers_Jason\powersenv\lib\site-packages\pytrigno-0.1-py3.5.egg\pytrigno.py", line 96, in read
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:/Users/emg/Documents/powers_Jason/pytrigno/examples/check_trigno.py", line 65, in <module>
    check_emg(args.host)
  File "c:/Users/emg/Documents/powers_Jason/pytrigno/examples/check_trigno.py", line 31, in check_emg
    data = dev.read()
  File "c:\Users\emg\Documents\powers_Jason\powersenv\lib\site-packages\pytrigno-0.1-py3.5.egg\pytrigno.py", line 220, in read
  File "c:\Users\emg\Documents\powers_Jason\powersenv\lib\site-packages\pytrigno-0.1-py3.5.egg\pytrigno.py", line 100, in read
OSError: Device disconnected.

**Comment by jdekarske** [comment]
If you can recall or take a guess on the firmware version(s), I can rollback and try again.

**Comment by ixjlyons** [comment]
I have no idea, but I suspect the firmware version shouldn't matter too much assuming TCU appears to be communicating with the device correctly. The library really only interfaces with TCU, so possibly if you have a newish version of that, Delsys could've changed the protocol a bit. Doesn't look like it to me though: https://www.delsys.com/downloads/USERSGUIDE/trigno/sdk.pdf

We know that if the test script has gotten to the point of attempting to read data from the device, which means it got through the initialization which involves connecting to the two ports and reading some initial data from the command port.

I honestly don't know why it's timing out. A couple things I can think of to help figure it out:

- Verify that the sensors are already paired and turned on, so before you run the script you can see the sensor(s) are there and powered on. I think if the sensors aren't on, TCU will just stream zeros, but I'm not sure what happens if they're not paired.
- Check visually that TCU responds right after you run the script. The status indicator for the sensor(s) should change to indicate they're actively being read from.
- Either use the Python debugger `pdb` or hack on a copy of the `pytrigno.py` module to determine if *any* of the `recv()` calls here are succeeding or if the very first one is timing out:

https://github.com/axopy/pytrigno/blob/2b2d8d5590935b7c5af81bcc2ced17ef2adf14e9/pytrigno.py#L96

- See if changing the timeout here to, say, 5 seconds helps:

https://github.com/axopy/pytrigno/blob/2b2d8d5590935b7c5af81bcc2ced17ef2adf14e9/pytrigno.py#L61-L62

**Comment by jdekarske** [comment]
TLDR: both sockets needed a longer timeout

- Sensors are paired and show live
- TCU Status changes to 'Remote (1 user)'
- Getting expected data from `recv()` when I step through `pytrigno.py` from `check_trigno.py` and if I put a break before stopping data collection.

After playing a bit, seems TCU isn't happy with the 'STOP' command:  
```
File "c:\Users\Jason\Documents\Git\partialPowersEMG\pytrigno\pytrigno.py", line 111, in stop
    self._send_cmd('STOP')
  File "c:\Users\Jason\Documents\Git\partialPowersEMG\pytrigno\pytrigno.py", line 125, in _send_cmd
    resp = self._comm_socket.recv(128)
socket.timeout: timed out
```

`warning: TrignoDaq command failed: b'CANNOT COMPLETE\r\n\r\n'`

From the documentation:

> 6.4 Control Command Replies
The server will respond “CANNOT COMPLETE” to a command which is valid, but cannot be completed due to the current state of the system (for example, a configuration command issued while data is streaming).

Seems like its the command socket. I changed the timeout here, in addition to , to 3 and `check_trigno.py` seems fine:
https://github.com/axopy/pytrigno/blob/2b2d8d5590935b7c5af81bcc2ced17ef2adf14e9/pytrigno.py#L56-L57

```
STOP b'OK\r\n\r\n'
START b'OK\r\n\r\n'
STOP b'OK\r\n\r\n'
START b'OK\r\n\r\n'
STOP b'OK\r\n\r\n'
```

I might need to play with it a bit more to check consistency, I'll follow up. Would you like a PR?

**Comment by ixjlyons** [comment]
Awesome. I don't recall if there was a reason I set it to 2 seconds aside from it seeming reasonable and working for me all this time.

> I might need to play with it a bit more to check consistency, I'll follow up. Would you like a PR?

:+1: I don't have access to the hardware anymore, so any improvements are welcome.
